### ISSUE TITLE
thumbnail with caching does not show foreground window in some situations

### ISSUE THREAD

**Issue by scottkosty** [mzur/gnome-shell-wsmatrix#33]
I think the cache is correctly regenerated when a new window is created on a workspace, but the regeneration does not seem to be regenerated when the ordering of the windows changes (e.g., I use alt-tab to bring a window to the foreground).

There might be some useful discussion on this issue at #31.

**Comment by ebeem** [comment]
Thanks for opening the issue, I am copying my observations here for reference


I did some investigation, this seems to be a GNOME issue and not related to caching.
steps to reproduce:
```
1- set timeout of workspace popup to 5000ms so you can easily notice the bug.
2- open a small window on the workspace (like nautilus)
3- open a video on the same workspace (like VLC)
4- show the workspace switcher (you can go to another workspace and back to your video workspace)
at this stage, you will notice that the video is playing well in the workspace thumbnail
5- alt-tab to switch application to your application in step#2 and show workspace switcher again
at this stage, you will notice the window hasn't been rendered in the thumbnail, but the video is still being animated
6- move the window at step#2 to another workspace or monitor, then bring it back to the original workspace or monitor
at this stage, you will notice that the window at step#2 has been rendered above the window at step#3 which is the expected behavior.
```

it seems like the GNOME team didn't handle all cases well in thumbnail preview.
in file `workspaceThumbnail.js`


```Javascript
        this._windows = [];
        this._allWindows = [];
        this._minimizedChangedIds = [];
        for (let i = 0; i < windows.length; i++) {
            let minimizedChangedId =
                windows[i].meta_window.connect('notify::minimized',
                                               this._updateMinimized.bind(this));
            this._allWindows.push(windows[i].meta_window);
            this._minimizedChangedIds.push(minimizedChangedId);

            if (this._isMyWindow(windows[i]) && this._isOverviewWindow(windows[i])) {
                this._addWindowClone(windows[i]);
            }
        }

        // Track window changes
        this._windowAddedId = this.metaWorkspace.connect('window-added',
                                                          this._windowAdded.bind(this));
        this._windowRemovedId = this.metaWorkspace.connect('window-removed',
                                                           this._windowRemoved.bind(this));
        this._windowEnteredMonitorId = global.display.connect('window-entered-monitor',
                                                              this._windowEnteredMonitor.bind(this));
        this._windowLeftMonitorId = global.display.connect('window-left-monitor',
                                                           this._windowLeftMonitor.bind(this));
```

looks like the thumbnail handles most changes in windows including minimize/add/remove window and left/entered monitors, but there's no window focus listener, which makes the thumbnail not getting updated if a window has been focused/unfocused.

I think we can make a workaround for this in the extension, but this is probably a GNOME bug and should be fixed by the GNOME team.

**Comment by mzur** [comment]
Thanks @ebeem. I second your opinion that this should be fixed in GNOME and not by Workspace Matrix. I'll leave this open anyway; perhaps one of us can find the time to make a PR to GNOME.

**Comment by mzur** [comment]
We should test if this is fixed in GNOME 3.36 ( #97 ).

**Comment by ebeem** [comment]
tested, the problem exists in GNOME 3.36
